proposals/tabs_query_filter_document_id.md

```diff
@@ -0,0 +1,126 @@
+propose new APIs and use this template.  Each section includes instructions on
+what to include.  Delete the instructions for a given section once it's filled
+out.  Remove this section once the template is filled out.
+
+**Summary**
+
+Extend `tabs.query()` first (filter) argument with `documentId` attribute.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** 
+
+**Created:** 2024-04-08
+
+**Related Issues:**
+https://github.com/w3c/webextensions/issues/553
+https://issues.chromium.org/issues/326749718
+https://chromium-review.googlesource.com/c/chromium/src/+/5318738
+
+## Motivation
+
+### Objective
+
+Extension interact with tabs in an explicitly asynchroneous manner, so every
+interaction is fundmentally prone to race conditions and
+time-of-check-time-of-use bugs. Currently extension may query tabs based on
+only incidental identifiers like `windowId`, `index`, and if extension has
+powerful `"tabs"` permission, then also `title` and `url`. Such checks are
+inherintly cumbersome for the extension developer, may lead to extension
+wrongfully assuming that a page navigation has occured when the page just
+changed its title or url programmatcally or extension may fail to detect page
+reload.
+This proposal aims to permit extension to query tabs directly by document id
+which is garanteed to persist across tab moves and guaranteed to change on page
+loads. This filter should be accessible without `"tabs"` permission to enable
+extensions to run without `"tabs"` permission.
+
+#### Use Cases
+
+Any use cases which use `tabs.query()` in cases when document navigation or
+load is likely.
+
+### Known Consumers
+
+Dark Reader would benefit from better availability of `documentId`. Dark Reader
+relies on `documentId` extensively on Chromium MV3 and emulates it if it is not
+availabke natively (Firefox, Chromium MV2) by creating an identifier in content
+script world and passing it everywhere.
+
+## Specification
+
+### Schema
+
+Extend `QueryInfo` argument passed to
+`tabs.query(queryInfo: QueryInfo, callback?: function)` with `documentId`
+which is a `string`.
+
+### Behavior
+
+1. If `documentId` is omitted, then `tabs.query()` behaves as before.
+2. If `documentId` is provided to `tabs.query()`, it is validated as any
+   other filter attribute and if provided value is malformed (not a `string`
+   or can not be parsed into proper document id), an error is thrown. If the
+   format of id is valid, but browser does not have a corresponding tab, then
+   `tabs.query()` succeeds with an empty response, since the id may represent
+   a tab not currently visible in the tab strip or a tab belonging to past
+   browsing session.
+3. If `documentId` is provided and valid, then `tabs.query()` returns only the
+   tab which hosts the corresponding document in a top-level context and
+   matching all other filters. (In particular, if `documentId` corresponds to
```

Comment:
I'm curious about this (the subframe behavior).  It seems like being able to query which tab holds a document could be useful, independent of if it's a main frame or subframe (e.g., for querying which tab holds a particular frame that the extension may have injected into via a scripting API).

Curious for other folks' thoughts on this.  Should we limit this to top frames?